lib/helpers.js

```diff
@@ -166,8 +166,26 @@ const getJavaHome = _.memoize(async function getJavaHome () {
 });
 
 const getJavaForOs = _.memoize(async function getJavaForOs () {
-  return path.resolve(await getJavaHome(), 'bin',
-    `java${system.isWindows() ? '.exe' : ''}`);
+  let javaHome;
+  let errMsg;
+  try {
+    javaHome = await getJavaHome();
+  } catch (err) {
+    errMsg = err.message;
+  }
+  const executableName = `java${system.isWindows() ? '.exe' : ''}`;
+  if (javaHome) {
+    const resultPath = path.resolve(javaHome, 'bin', executableName);
+    if (await fs.exists(resultPath)) {
+      return resultPath;
+    }
+  }
+  try {
+    return await fs.which(executableName);
```

Comment:
should we generalize this for other binaries, not just java?

Reply:
which binaries exactly you'd like to generalize?

Reply:
i'm not sure, don't we do similar logic currently for binaries inside ANDROID_HOME, and, e.g., jarsigner?

Reply:
jarsigner is only used as a fallback if apksigner fails for some reason, so I would rather leave it as is for now